### PR TITLE
DCGEO-138: ACO authorization changes based on "Data" attribute

### PR DESCRIPTION
**Pull request by KyRobbins**
### Fixes [DCGEO-138](https://jira.cms.gov/browse/DCGEO-138)

Updating the `requests` module to access control on the `type` of FHIR resource requested based on the `Data` permissions defined per ACO Model in the configuration file.  If a model has `adjudicated` access, they will obtain permissions to request `Patient, Coverage and EOB` resources.  If the model has `pre-adjudicated` access, they will obtain permissions to request `Claim and ClaimResponse` resources.  Additionally, the default FHIR resource types have been updated to match the above logic.

### Proposed Changes

Update to allow data-source specification of resources and make ACO level configurations for authorized data types

### Change Details

- Add additional valid Resource types
- Make Resource types indicate valid data source types
- Update validation checks to include authorization check for data type

### Security Implications

ACOs should only have access to their authorized claim data and FHIR Resource types

- [ ] new software dependencies

- [ ] security controls or supporting software altered

- [ ] new data stored or transmitted

- [ ] security checklist is completed for this change

- [ ] requires more information or team discussion to evaluate security implications

- [ ] no PHI/PII is affected by this change

### CODE REVIEW

File: bcda/api/requests.go [hunk 1]
```diff
@@ -34,6 +34,11 @@ import (
 	"github.com/CMSgov/bcda-app/log"
 )
 
+type ResourceType struct {
```
Comment (ericbuckley) [inline bcda/api/requests.go:37]: can we add a comment block here describing the purpose of this struct.

File: bcda/api/requests.go [hunk 2]
```diff
@@ -280,13 +282,9 @@ func (h *Handler) bulkRequest(w http.ResponseWriter, r *http.Request, reqType se
...
 
-	resourceTypes := rp.ResourceTypes
-	// If caller does not supply resource types, we default to all supported resource types
-	if len(resourceTypes) == 0 {
-		resourceTypes = []string{"Patient", "ExplanationOfBenefit", "Coverage"}
-	}
+	resourceTypes := h.getResourceTypes(rp, ad.CMSID)
 
-	if err = h.validateRequest(resourceTypes); err != nil {
+	if err = h.validateRequest(resourceTypes, ad.CMSID); err != nil { //TODO: Validate type of claim & resource data for user
```
Comment (ericbuckley) [inline bcda/api/requests.go:287]: is the TODO still needed?

File: bcda/api/requests.go [hunk 3]
```diff
@@ -424,17 +422,61 @@ func (h *Handler) bulkRequest(w http.ResponseWriter, r *http.Request, reqType se
...
+func (h *Handler) authorizedResourceAccess(resource ResourceType, cmsId string) bool {
+	if cfg, ok := h.Svc.GetACOConfigForId(cmsId); ok {
+		return (resource.Adjudicated && contains(cfg.Data, "adjudicated")) ||
+			(resource.PreAdjudicated && contains(cfg.Data, "preadjudicated"))
+	}
+
+	return false
+}
+
+func contains(array []string, str string) bool {
```
Comment (ericbuckley) [inline bcda/api/requests.go:470]: can we use `ContainsString` instead?  https://github.com/CMSgov/bcda-app/blob/master/bcda/utils/common.go#L63

File: bcda/api/v1/api.go [hunk 4]
```diff
@@ -23,7 +23,14 @@ import (
 var h *api.Handler
 
 func init() {
-	h = api.NewHandler([]string{"Patient", "Coverage", "ExplanationOfBenefit", "Observation"}, "/v1/fhir", "v1")
+	resources := make(map[string]api.ResourceType, 6)
+	resources["Patient"] = api.ResourceType{Adjudicated: true}
+	resources["Coverage"] = api.ResourceType{Adjudicated: true}
+	resources["ExplanationOfBenefit"] = api.ResourceType{Adjudicated: true}
+	resources["Observation"] = api.ResourceType{Adjudicated: true}
+	resources["Claim"] = api.ResourceType{Adjudicated: false, PreAdjudicated: true}
```
Comment (ericbuckley) [inline bcda/api/v1/api.go:31]: Claim and ClaimResponse shouldn't be options for v1, BFD will only support pre-adj data in v2

File: bcda/service/service.go [hunk 5]
```diff
@@ -456,6 +458,16 @@ func (s *service) GetJobPriority(acoID string, resourceType string, sinceParam b
 	return priority
 }
 
+func (s *service) GetACOConfigForId(cmsId string) (*ACOConfig, bool) {
```
Comment (ericbuckley) [inline bcda/service/service.go:461]: thanks for this, I think other parts of the codebase can reuse this function.  mind adding a docstring?

**Review by ericbuckley — COMMENTED** [review]
@KyRobbins just a few comments.  Logic seems pretty sound, I think we just need to clean up a few things

**Review by ericbuckley — APPROVED** [review]
LGTM

**Review by brianburton — DISMISSED** [review]
LGTM

**Review by ericbuckley — DISMISSED** [review]

**Review by ericbuckley — APPROVED** [review]
LGTM

**Review by brianburton — APPROVED** [review]
LGTM

File: bcda/api/requests.go [hunk 6]
```diff
@@ -514,17 +515,55 @@ func (h *Handler) bulkRequest(w http.ResponseWriter, r *http.Request, reqType se
...
+			}
+		}
+	}
+
+	return resourceTypes
+}
+
+func (h *Handler) validateRequest(resourceTypes []string, cmsID string) error {
 
 	for _, resourceType := range resourceTypes {
```
Comment (katyasoup) [inline bcda/api/requests.go:539]: `resourceType` is a bit confusing here, as it's referring to a string, rather than the ResourceType struct. Could we maybe use something like `resourceString` instead (and perhaps `resourceType` in place of `resource`)? Or on the flip side, maybe `ResourceType` could be `ResourceAccess`/`ResourceData` (or similar)?

File: bcda/api/requests.go [hunk 7]
```diff
@@ -514,17 +515,55 @@ func (h *Handler) bulkRequest(w http.ResponseWriter, r *http.Request, reqType se
...
-		if _, ok := h.supportedResources[resourceType]; !ok {
-			return fmt.Errorf("invalid resource type %s. Supported types %s", resourceType, h.supportedResources)
+		resource, ok := h.supportedResources[resourceType]
+
+		if !ok {
+			resources := make([]string, 0, len(h.supportedResources))
+			for k := range h.supportedResources {
```
Comment (katyasoup) [inline bcda/api/requests.go:544]: Could we move this out (maybe to the `init()`) and avoid having to run this extra loop for every resource param? Having the slice of resource strings available might be handy elsewhere, too.

File: bcda/api/requests.go [hunk 8]
```diff
@@ -514,17 +515,55 @@ func (h *Handler) bulkRequest(w http.ResponseWriter, r *http.Request, reqType se
 	}
 }
 
-func (h *Handler) validateRequest(resourceTypes []string) error {
+func (h *Handler) getResourceTypes(parameters middleware.RequestParameters, cmsID string) []string {
+	resourceTypes := parameters.ResourceTypes
```
Comment (katyasoup) [inline bcda/api/requests.go:519]: see below re: resourceType vs. ResourceType confusion

**Review by katyasoup — COMMENTED** [review]
Looking good! A few clarifying Qs in `requests.go`

File: bcda/api/requests.go (hunk 8 above)
Comment (ericbuckley) [reply bcda/api/requests.go:519]: @katyasoup yeah I agree, the `resource type` seems to have a couple meanings and can be confusing depending on the context.  I was thinking about changing up the labels here to the following, what do you think?

resource type = FHIR resource type (eg Patient, EOB, Claim, Coverage, etc)
data type = upstream source data (eg adjudicated vs pre-adjudicated)

Comment (katyasoup) [reply bcda/api/requests.go:519]: That sounds great! Then we'd have a map of `[resourceType]dataType` which I think is much easier to follow

**Review by katyasoup — DISMISSED** [review]
LGTM! Nice work!

**Review by brianburton — APPROVED** [review]

**Review by ericbuckley — APPROVED** [review]

File: bcda/api/requests_test.go [hunk 9]
```diff
@@ -219,7 +232,13 @@ func (s *RequestsTestSuite) TestRunoutDisabled() {
...
+	resourceMap := map[string]DataType{
+		"Patient":              {Adjudicated: true},
+		"Coverage":             {Adjudicated: true},
+		"ExplanationOfBenefit": {Adjudicated: true},
+	}
+
+	h := newHandler(resourceMap, "/v1/fhir", "v1", s.db)
 
 	// Use a mock to ensure that this test does not generate artifacts in the queue for other tests
 	enqueuer := &queueing.MockEnqueuer{}
```
Comment (CosmicSyntax) [inline bcda/api/requests_test.go:244]: Should we add a test for validating adjudicated? Please let me know if this is already done somewhere and I missed it.

**Review by CosmicSyntax — COMMENTED** [review]
This looks great. Just one comment on testing in request_test.go.

File: bcda/api/requests_test.go (hunk 9 above)
Comment (ericbuckley) [reply bcda/api/requests_test.go:244]: Do you mean a test for validating pre-adjudicated?  From the looks of it, the `TestRequests` function is validating v1 adjudicated requests.

Comment (CosmicSyntax) [reply bcda/api/requests_test.go:244]: Is adjudicated always set to true, no matter the ACO model?

Comment (ericbuckley) [reply bcda/api/requests_test.go:244]: No, we're having it be explicitly set, thus the need for having the ACO config models updated across the environments before this is merged.  If an ACO model does not have adjudicated listed in the configuration, they will get a permissions error when making a request for any of the CCW BFD data.

Comment (CosmicSyntax) [reply bcda/api/requests_test.go:244]: So I believe all request tests are done using this dummy ACO: "ba21d24d-cd96-4d7d-a691-b0e8c88e67a5." Which has a cms_id of "A0002." Should we add another dummy model and test for a negative where adjudicated is not set?

Comment (ericbuckley) [reply bcda/api/requests_test.go:244]: Ah ha, I see what you're saying now.  Yes I agree, we should ideally have tests for ACOs testing all 4 possibly states (or at least the first 3).
```
1. [adjudicated, pre-adjudicated]
2. [adjudicated]
3. [pre-adjudicated]
4. []
```

File: bcda/api/requests_test.go [hunk 10]
```diff
@@ -216,10 +231,116 @@ func (s *RequestsTestSuite) TestRunoutDisabled() {
 	s.Contains(string(body), "Invalid group ID")
 }
 
+func (s *RequestsTestSuite) TestDataTypeAuthorization() {
```
Comment (ericbuckley) [inline bcda/api/requests_test.go:234]: thanks for adding this test case @KyRobbins!

**Review by ericbuckley — APPROVED** [review]
🎉

File: bcda/api/requests_test.go [hunk 11]
```diff
@@ -229,6 +231,107 @@ func (s *RequestsTestSuite) TestRunoutDisabled() {
...
+		{"Auth Pre-Adj, Request Adj", "C0000", []string{"Patient"}, http.StatusBadRequest},
+		{"Auth Pre-Adj, Request Pre-Adj", "C0000", []string{"Claim"}, http.StatusAccepted},
+		{"Auth None, Request Adj", "D0000", []string{"Patient"}, http.StatusBadRequest},
+		{"Auth None, Request Pre-Adj", "D0000", []string{"Claim"}, http.StatusBadRequest},
+	}
+
+	for _, test := range tests {
+		s.T().Run(test.name, func(t *testing.T) {
+			w := httptest.NewRecorder()
+			r, _ := http.NewRequest("GET", "http://bcda.ms.gov/api/v2/Group/$export", bytes.NewReader(jsonBytes))
```
Comment (CosmicSyntax) [inline bcda/api/requests_test.go:315]: If we are sending an empty body, we can leave this as nil. But not a big deal to change.

File: bcda/api/requests_test.go [hunk 12]
```diff
@@ -322,3 +425,47 @@ func (s *RequestsTestSuite) genASRequest() *http.Request {
 
 	return req.WithContext(ctx)
 }
+
+type MockService struct {
```
Comment (CosmicSyntax) [inline bcda/api/requests_test.go:429]: We usually auto-generate mocks using mockery. But we can create a separate ticket to do this later.

**Review by CosmicSyntax — APPROVED** [review]
🎉

**Review by ericbuckley — APPROVED** [review]

**Review by katyasoup — APPROVED** [review]
🎉 LGTM!

File: bcda/api/requests_test.go (hunk 12 above)
Comment (SMLuthi) [reply bcda/api/requests_test.go:429]: Actually, I think we should switch to using the mock service thats already in place to avoid stacking up tech debt. Look for usage of `mockSvc` in this file.

**Review by SMLuthi — CHANGES_REQUESTED** [review]

**Review by SMLuthi — APPROVED** [review]
Since Kyle is addressing the mock service stuff in his next PR ill just approve this!

https://github.com/CMSgov/bcda-app/pull/733/files#diff-7fc3cc46a29844127337e64bc5dbd5d50fbfa043073ae134ea665a8c07cedd9bR305
